services/galley/test/integration/API/MLS.hs

```diff
@@ -670,6 +705,134 @@ testCommitNotReferencingAllProposals = withSystemTempDirectory "mls" $ \tmp -> d
+        NonEmpty.tail (pClients creator) <> toList (pClients bob)
+
+  testSuccessfulCommit MessagingSetup {users = [bob], commit = addCommit, ..}
+
+  -- remove only first client of bob
+  (removalCommit, _mbWelcome) <- liftIO $ setupRemoveCommit tmp creator "group" "group" [(NE.head (pClients bob))]
+
+  err <-
+    responseJsonError
+      =<< postMessage (qUnqualified (pUserId creator)) removalCommit
+        <!! statusCode === const 409
+
+  liftIO $ Wai.label err @?= "mls-client-mismatch"
+
+testUserRemovesThemselvesFromConv :: TestM ()
+testUserRemovesThemselvesFromConv = withSystemTempDirectory "mls" $ \tmp -> do
+  (creator, [bob]) <- withLastPrekeys $ setupParticipants tmp def [(2, LocalUser)]
+
+  -- create a group
+  (groupId, conversation) <- setupGroup tmp CreateConv creator "group"
+
+  -- add clients to it and get welcome message
+  (addCommit, welcome) <-
+    liftIO $
+      setupCommit tmp creator "group" "group" $
+        NonEmpty.tail (pClients creator) <> toList (pClients bob)
+
+  testSuccessfulCommit MessagingSetup {users = [bob], commit = addCommit, ..}
+
+  -- FUTUREWORK: create commit as bob, when the openmls library supports removing own clients
+  (removalCommit, _mbWelcome) <- liftIO $ setupRemoveCommit tmp creator "group" "group" (pClients bob)
+
+  -- bob tries to leave the conversation by removng all its clients
+  err <-
+    responseJsonError
+      =<< postMessage (qUnqualified (pUserId bob)) removalCommit
+        <!! statusCode === const 409
+
+  liftIO $ Wai.label err @?= "mls-self-removal-not-allowed"
+
+testRemoveDeletedClient :: Bool -> TestM ()
+testRemoveDeletedClient deleteClientBefore = withSystemTempDirectory "mls" $ \tmp -> do
+  (creator, [bob, dee]) <- withLastPrekeys $ setupParticipants tmp def [(2, LocalUser), (1, LocalUser)]
+
+  -- create a group
+  (groupId, conversation) <- setupGroup tmp CreateConv creator "group"
+
+  -- add clients to it and get welcome message
+  (addCommit, welcome) <-
+    liftIO $
+      setupCommit tmp creator "group" "group" $
+        NonEmpty.tail (pClients creator) <> toList (pClients bob) <> toList (pClients dee)
+
+  testSuccessfulCommit MessagingSetup {users = [bob, dee], commit = addCommit, ..}
+
+  let (_bobClient1, bobClient2) = assertTwo (toList (pClients bob))
+
+  when deleteClientBefore $
+    deleteClient (qUnqualified (pUserId bob)) (snd bobClient2) (Just defPassword)
+      !!! statusCode === const 200
+
+  (removalCommit, _mbWelcome) <- liftIO $ setupRemoveCommit tmp creator "group" "group" [bobClient2]
+
+  -- dee (which is not an admin) commits removal of bob's deleted client
+  let doCommitRemoval = postMessage (qUnqualified (pUserId dee)) removalCommit
+
+  if deleteClientBefore
+    then do
+      events :: [Event] <-
+        responseJsonError
+          =<< doCommitRemoval
+            <!! statusCode === const 201
+      liftIO $ assertEqual "no conversation events from removing client" [] events
```

Comment:
Shouldn't the error message be the opposite, i.e., something like "a non-admin received conversation events when removing a client"?

Reply:
Right, I've updated the message